src/proxy/ProxyManager.ts

```diff
@@ -79,140 +78,151 @@ export class ProxyManager {
-        this.service[ProxyManager.ATOMICLONG_SERVICE] = AtomicLongProxy;
         this.service[ProxyManager.FLAKEID_SERVICE] = FlakeIdGeneratorProxy;
         this.service[ProxyManager.PNCOUNTER_SERVICE] = PNCounterProxy;
         this.service[ProxyManager.RELIABLETOPIC_SERVICE] = ReliableTopicProxy;
     }
 
     public getOrCreateProxy(name: string, serviceName: string, createAtServer = true): Promise<DistributedObject> {
         const fullName = serviceName + name;
-        if (this.proxies[fullName]) {
-            return this.proxies[fullName];
+        if (this.proxies.has(fullName)) {
+            return this.proxies.get(fullName);
         }
 
         const deferred = DeferredPromise<DistributedObject>();
-        let newProxy: DistributedObject;
-        if (serviceName === ProxyManager.MAP_SERVICE && this.client.getConfig().getNearCacheConfig(name)) {
-            newProxy = new NearCachedMapProxy(this.client, serviceName, name);
-        } else if (serviceName === ProxyManager.RELIABLETOPIC_SERVICE) {
-            newProxy = new ReliableTopicProxy(this.client, serviceName, name);
-            if (createAtServer) {
-                (newProxy as ReliableTopicProxy<any>).setRingbuffer().then(() => {
-                    return this.createProxy(newProxy);
-                }).then(function (): void {
-                    deferred.resolve(newProxy);
-                });
-            }
-            this.proxies[fullName] = deferred.promise;
-            return deferred.promise;
+        this.proxies.set(fullName, deferred.promise);
+
+        let createProxyPromise: Promise<any>;
+        if (createAtServer) {
+            createProxyPromise = this.createProxy(name, serviceName);
         } else {
-            newProxy = new this.service[serviceName](this.client, serviceName, name);
+            createProxyPromise = Promise.resolve();
         }
 
-        if (createAtServer) {
-            this.createProxy(newProxy).then(function (): void {
-                deferred.resolve(newProxy);
+        createProxyPromise
+            .then(() => {
+                return this.initializeLocalProxy(name, serviceName, createAtServer);
+            })
+            .then((localProxy) => {
+                deferred.resolve(localProxy);
+            })
+            .catch((error) => {
+                this.proxies.delete(fullName);
+                deferred.reject(error);
             });
-        }
 
-        this.proxies[fullName] = deferred.promise;
         return deferred.promise;
     }
 
-    destroyProxy(name: string, serviceName: string): Promise<void> {
-        delete this.proxies[serviceName + name];
+    public createDistributedObjectsOnCluster(): Promise<void> {
+        const proxyEntries = new Array<[string, string]>();
+        for (const namespace of Array.from(this.proxies.keys())) {
+            const promise = this.proxies.get(namespace);
+            if (promise.isFulfilled()) {
```

Comment:
You are right. We were using `serviceName + name` string as the keys of proxies so there were no way to extract them from the concatenated string. That's why I did a check like this but this is a different behavior as you said. I added a separator between service name and name so that we can extract them from the namespace string so there is no need for a check like this